mdnshosts/main.c

```diff
@@ -0,0 +1,684 @@
+		warn("pledge");
+		return(0);
+	}
+
+	/* Wait until the renamer has started up. */
+
+	if (0 == (c = read_token(wfd, "renamer up", NULL))) {
+		warnx("process down: renamer");
+		return(0);
+	} else if (-1 == c)
+		return(0);
+
+	/*
+	 * Main loop: wait until we receive an update on our hosts, then
+	 * flush any changes to the file hosts.
+	 * Re-create this every time.
+	 */
+
+	for (;;) {
+		if (0 == (c = read_token(rfd, "update", &tok))) {
+			warnx("writer: updater has exited");
+			rc = 1;
+			break;
+		} else if (c < 0)
+			break;
+
+		change = 0;
+
+		/* Process addition or deletion. */
+
+		if (1 == tok) {
+			if ( ! read_str(rfd, "name", &name))
+				break;
+
+			if (NULL != (ent = db_lookup(&db, name))) {
+				change = 1 == ent->refs;
+				db_unref(ent);
+			} else
+				warnx("unknown service: %s", name);
+
+			free(name);
+			name = NULL;
+		} else if (2 == tok) {
+			if ( ! read_str(rfd, "name", &name))
+				break;
+			if ( ! read_buf(rfd, "addr", &addr, sizeof(addr)))
+				break;
+			if ( ! read_str(rfd, "target", &target))
+				break;
+
+			if (NULL != (ent = db_lookup(&db, name))) {
+				free(name);
+				change = 0 == ent->refs;
+				db_ref(ent, target, &addr);
+			} else {
+				change = 1;
+				db_add(&db, name, target, &addr);
+			}
+			name = target = NULL;
+		} else 
+			err(EXIT_FAILURE, "unknown token");
+
+		if ( ! change)
+			continue;
+
+		if (NULL == (f = fopen("hosts", "a"))) {
+			warn("/etc/mdns/hosts");
+			break;
+		}
+
+		fprintf(f, "127.0.0.1 localhost\n");
+		fprintf(f, "::1 localhost\n");
+		for (i = 0; i < db.hostsz; i++) {
+			fprintf(f, "%s %s\n", 
+				inet_ntoa(db.hosts[i].addr), 
+				db.hosts[i].target);
+		}
+		fclose(f);
+
+		if (-1 == chmod("hosts", 0644)) {
+			warn("/etc/mdns/hosts");
+			break;
+		}
+
+		warnx("writer: created /etc/mdns/hosts");
+		if ( ! write_token(wfd, "writer req", 1))
+			break;
+		if ( ! read_token(wfd, "writer resp", NULL))
+			break;
+		warnx("writer: changes flushed");
+	}
+
+	free(target);
+	free(name);
+
+	for (i = 0; i < db.hostsz; i++) {
+		free(db.hosts[i].name);
+		free(db.hosts[i].target);
+	}
+
+	free(db.hosts);
+	return(rc);
+}
+
+static int
+proc_renamer(int fd)
+{
+	int	 c, rc = 0;
+
+	warnx("renamer: start");
+
+	/* 
+	 * Make us safe: chroot in /etc and pledge.
+	 * We need to use rename(2) (and thus cpath) in this because
+	 * /etc/hosts needs to have sane contents at all times.
+	 */
+
+	if (-1 == chroot("/etc")) {
+		warn("chroot");
+		return(0);
+	} else if (-1 == chdir("/")) {
+		warnx("chdir");
+		return(0);
+	} else if (-1 == pledge("stdio rpath cpath", NULL)) {
+		warn("pledge");
+		return(0);
+	}
+
+	warnx("renamer: /etc/hosts <-> /etc/hosts/mdns.save");
+
+	/* Back up our current /etc/hosts. */
+
+	if (-1 == link("hosts", "hosts.mdns.save")) {
+		warn("link: /etc/hosts, /etc/hosts.mdns.save");
+		return(0);
+	}
+
+	warnx("renamer: notifying reader");
+
+	/* Notify the writer that we're ready. */
+
+	if ( ! write_token(fd, "reader up", 1))
```

Comment:
I won't bother with the if (-1 == foo) since this is your work, but I'd ask if possible, to keep the ! together, like if (!write_token()) or (write_token() == 0) or (0 == write_token()).